### PR TITLE
Be more like a library to support mobile

### PR DESCRIPTION
**Pull request by nbrownus**
The goal here is for `nebula` to behave more like a library where an implementation gets greater control over the process. This is needed to enable mobile clients like ios or android.

The command channel will be reworked in another PR. The set that exists today is the minimum required for mobile.

### CODE REVIEW

File: udp_android.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package nebula
+
+// Darwin support is primarily implemented in udp_generic, besides NewListenConfig
```
Comment (forfuncsake) [inline udp_android.go:3]: This comment doesn't appear relevant to the `udp_android.go` file.

File: tun_ios.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+	return nil
+}
+
+func (c *Tun) WriteRaw(b []byte) error {
+	_, err := c.Write(b)
+	return err
+}
+
+// The following is hoisted up from water, we do this so we can inject our own fd on iOS
+type tunReadCloser struct {
```
[the remark this answers is not in the document]
Comment (nbrownus) [reply tun_ios.go:49]: This code comes directly from water. The real "correct" fix would be to support providing a file descriptor in water itself. That is outside the scope of this change though.

File: udp_android.go (hunk 1 above)
Comment (nbrownus) [reply udp_android.go:3]: Copy pasta!

File: main.go [hunk 3]
```diff
@@ -103,19 +102,28 @@ func Main(configPath string, configTest bool, buildVersion string) {
 
 	var tun *Tun
 	if !configTest {
-		config.CatchHUP()
```
[the remark this answers is not in the document]
Comment (nbrownus) [reply main.go:106]: Negative, thanks for looking!

File: cmd/nebula/main.go [hunk 4]
```diff
@@ -3,6 +3,7 @@ package main
 import (
 	"flag"
 	"fmt"
+	"github.com/sirupsen/logrus"
 	"os"
 
 	"github.com/slackhq/nebula"
```
[the remark this answers is not in the document]
Comment (nbrownus) [reply cmd/nebula/main.go:9]: Yeah it's a `go fmt` vs `goimports` thing. Guess it's time I start using the latter as well.

[the remark this answers is not in the document]
Comment (nbrownus) [reply cmd/nebula/main.go:9]: Talked with @wadey and we agreed this is a great change to make in an isolated PR and then rig up a check to make sure `goimports` was used on each PR.

File: logger.go [hunk 5]
```diff
@@ -0,0 +1,27 @@
+package nebula
+
+import (
+	"github.com/sirupsen/logrus"
+)
+
+type ContextualError struct {
```
Comment (wadey) [inline logger.go:7]: This should also implement the `Unwrap() error` method expected on error wrapping types.

File: main.go [hunk 6]
```diff
@@ -392,5 +433,9 @@ func shutdownBlock(ifce *Interface) {
 	ifce.hostMap.Unlock()
 
 	l.WithField("signal", sig).Info("Goodbye")
-	os.Exit(0)
+	if cmd.Callback != nil {
+		select {
+		case cmd.Callback <- nil:
+		}
+	}
```
Comment (wadey) [inline main.go:440]: You don't need the nil check here, and if you want the select to be non-blocking you need to add a `default:` case
